clarity/src/vm/ast/parser_v2/mod.rs

```diff
@@ -0,0 +1,3227 @@
+                let mut e = PreSymbolicExpression::atom(name);
+                e.span = token.span;
+                Some(e)
+            }
+            Token::Placeholder(s) => {
+                let mut e = PreSymbolicExpression::placeholder(s.to_string());
+                e.span = token.span;
+                Some(e)
+            }
+            Token::Comment(comment) => {
+                let mut e = PreSymbolicExpression::comment(comment.to_string());
+                e.span = token.span;
+                Some(e)
+            }
+            Token::Eof => None,
+            _ => None, // Other tokens should be dealt with by the caller
+        };
+
+        Ok(node)
+    }
+
+    pub fn parse_node_or_eof(&mut self) -> ParseResult<Option<PreSymbolicExpression>> {
+        loop {
+            match self.parse_node()? {
+                Some(node) => break Ok(Some(node)),
+                None => {
+                    let token = self.tokens[self.next_token - 1].clone();
+                    match token.token {
+                        Token::Eof => break Ok(None),
+                        _ => {
+                            // Report an error, then skip this token
+                            self.add_diagnostic(
+                                ParseErrors::UnexpectedToken(token.token),
+                                token.span,
+                            )?;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn parse(&mut self) -> ParseResult<Vec<PreSymbolicExpression>> {
+        let mut nodes = vec![];
+
+        loop {
+            match self.parse_node_or_eof()? {
+                Some(node) => {
+                    nodes.push(node);
+                }
+                None => break,
+            }
+        }
+        Ok(nodes)
+    }
+}
+
+pub fn parse(input: &str) -> ParseResult<Vec<PreSymbolicExpression>> {
+    let mut parser = match Parser::new(input, true) {
+        Ok(parser) => parser,
+        Err(e) => return Err(ParseError::new(e)),
+    };
+    let stmts = parser.parse()?;
+    if parser.success {
+        Ok(stmts)
+    } else {
+        let err = parser.diagnostics.remove(0);
+        Err(ParseError::new(err.e))
+    }
+}
+
+pub fn parse_collect_diagnostics(
+    input: &str,
+) -> (Vec<PreSymbolicExpression>, Vec<Diagnostic>, bool) {
+    // When not in fail_fast mode, Parser::new always returns Ok.
+    let mut parser = Parser::new(input, false).unwrap();
+
+    // When not in fail_fast mode, Parser::parse always returns Ok.
+    let stmts = parser.parse().unwrap();
+    let diagnostics = parser
+        .diagnostics
+        .iter()
+        .map(|e| Diagnostic {
+            level: e.e.level(),
+            message: e.e.message(),
+            spans: vec![e.span.clone()],
+            suggestion: None,
+        })
+        .collect();
+    (stmts, diagnostics, parser.success)
+}
+
+pub fn decode_hex(s: &str) -> Result<Vec<u8>, ParseIntError> {
```

Comment:
This duplicates `stacks_common::util::hash::hex_bytes`.  Can you use that method instead?

Reply:
Great! Yes, removed in 9324482ec.